### PR TITLE
Add scala 3 code demos for Scala for Java Programmers

### PR DESCRIPTION
**Pull request by bishabosha**
Update text a bit, however the whole thing reads still like it is for 2006 Java programmers.

### CODE REVIEW

File: _overviews/tutorials/scala-for-java-programmers.md [hunk 1]
```diff
@@ -159,85 +220,133 @@ types.
...
-        oncePerSecond(timeFlies)
-      }
-    }
+{% tabs callback-demo class=tabs-scala-version %}
+
+{% tab 'Scala 2' for=callback-demo %}
+```scala
+object Timer {
+  def oncePerSecond(callback: () => Unit): Unit = {
+    while (true) { callback(); Thread sleep 1000 }
```
Comment (julienrf) [inline _overviews/tutorials/scala-for-java-programmers.md:277]: I think there is an opportunity to change the code style to this:

```suggestion
    while (true) { callback(); Thread.sleep(1000) }
```

File: _overviews/tutorials/scala-for-java-programmers.md [hunk 2]
```diff
@@ -159,85 +220,133 @@ types.
...
+Here is a revised version of the timer
+program, passing an anonymous function to `oncePerSecond` instead of `timeFlies`:
+
+{% tabs callback-demo-refined class=tabs-scala-version %}
+
+{% tab 'Scala 2' for=callback-demo-refined %}
+```scala
+object TimerAnonymous {
+  def oncePerSecond(callback: () => Unit): Unit = {
+    while (true) { callback(); Thread sleep 1000 }
```
Comment (julienrf) [inline _overviews/tutorials/scala-for-java-programmers.md:322]:
```suggestion
    while (true) { callback(); Thread.sleep(1000) }
```

File: _overviews/tutorials/scala-for-java-programmers.md [hunk 3]
```diff
@@ -607,11 +969,41 @@ individual fields. The default implementation of `equals` is not
...
+
+{% tab 'Scala 2' for=equals-definition %}
+```scala
+class Date(y: Int, m: Int, d: Int) extends Ord {
+  // previous decls here
+
+  override def equals(that: Any): Boolean =
+    that.isInstanceOf[Date] && {
+      val o = that.asInstanceOf[Date]
+      o.day == day && o.month == month && o.year == year
```
Comment (julienrf) [inline _overviews/tutorials/scala-for-java-programmers.md:982]: Should we use pattern matching instead?

Comment (julienrf) [reply _overviews/tutorials/scala-for-java-programmers.md:982]: Eg

```suggestion
    that match {
      case date: Date => date.day == day && date.month == month && date.year == year
      case _ => false
    }
```

**Review by julienrf — COMMENTED** [review]
Awesome, Jamie! It not only adds Scala 3 code examples, but it also greatly improves the content of the document!

File: _overviews/tutorials/scala-for-java-programmers.md (hunk 3 above)
Comment (bishabosha) [reply _overviews/tutorials/scala-for-java-programmers.md:982]: im not a fan of keeping this section, but ill make the change

Comment (bishabosha) [reply _overviews/tutorials/scala-for-java-programmers.md]: I kept the comparison of `instanceof` and `isInstanceOf`, but instead say it is not idiomatic to use instance of in Scala.

File: _overviews/tutorials/scala-for-java-programmers.md (hunk 2 above)
Comment (bishabosha) [reply _overviews/tutorials/scala-for-java-programmers.md]: good catch, done

File: _overviews/tutorials/scala-for-java-programmers.md (hunk 1 above)
Comment (bishabosha) [reply _overviews/tutorials/scala-for-java-programmers.md]: done

**Review by julienrf — APPROVED** [review]
Thank you again Jamie!

**Pull request merged**